x-pack/plugins/fleet/server/services/actions/actions.ts

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { partition } from 'lodash';
+import { v4 as uuidV4 } from 'uuid';
+import type { ElasticsearchClient } from '@kbn/core-elasticsearch-server';
+import { fromKueryExpression, toElasticsearchQuery } from '@kbn/es-query';
+import type { estypes } from '@elastic/elasticsearch';
+
+import { FleetActionsError } from '../../../common/errors';
+import { AGENT_ACTIONS_INDEX, AGENT_ACTIONS_RESULTS_INDEX } from '../../../common';
+import { auditLoggingService } from '../audit_logging';
+
+import { appContextService } from '../app_context';
+
+import type { FleetAction, FleetActionResponse } from './types';
+
+export const createAction = async (
+  esClient: ElasticsearchClient,
+  action: Partial<FleetAction>
+): Promise<Partial<FleetAction>> => {
+  try {
+    const body = {
+      ...action,
+      action_id: action.action_id ?? uuidV4(),
+      '@timestamp': action['@timestamp'] ?? new Date().toISOString(),
+    };
+    await esClient.create(
+      {
+        index: AGENT_ACTIONS_INDEX,
+        id: body.action_id,
+        body,
+        refresh: 'wait_for',
+      },
+      {
+        meta: true,
+      }
+    );
+
+    auditLoggingService.writeCustomAuditLog({
+      message: `User created Fleet action [id=${action.action_id}]`,
+    });
+
+    return body;
+  } catch (createActionError) {
+    throw new FleetActionsError(
+      `Error creating action: ${createActionError.message}`,
+      createActionError
+    );
+  }
+};
+
+type BulkCreate = Array<{ create: { _index: string; _id: string } } | object>;
+
+export const bulkCreateActions = async (
+  esClient: ElasticsearchClient,
+  _actions: Array<Partial<FleetAction>>
+): Promise<
```

Comment:
same here: we should return an array of `FleetAction` documents (as long as the `bulk()` method returns the created documents - I don't remember)

Reply:
So the bulk method creation doesn't return created documents, only a set of objects that inform on successful and failed bulk operations, in the order of the documents given to the bulk method. So we've to filter out the given actions based on `errors` present in the response.